### PR TITLE
[FLINK-26520][table] Implement SEARCH operator in codegen

### PR DESCRIPTION
**Pull request by slinkydeveloper**
## What is the purpose of the change

This PR implements the `SEARCH` operator in the codegen, and removes the scalar implementation of `IN` and `NOT_IN`. Now every scalar `IN`/`NOT_IN` using a constant set is implemented through `SEARCH` (following Calcite's development on the topic https://issues.apache.org/jira/browse/CALCITE-4173) and plans will only have `SEARCH`

## Brief change log

* Remove `NOT_IN` from the plans, which was added only by the `ConvertToNotInOrInRule`. With this commit every scalar `IN`/`NOT_IN` using a constant set is converted to `SEARCH`, otherwise it's converted to a chain of disjunctions (see `RexUtil#expandSearch`)
* Now `GenerateUtils#generateLiteral` understands the Flink's internal type system. The conversion from Calcite's Comparable type system to Flink's internal type system is provided by a new function `RexLiteralUtil#toFlinkInternalValue`
* Add `SearchOperatorGen` to implement the `SEARCH` operator starting from the previous `generateIn` function in `ScalarOperatorGens`.

## Verifying this change

Existing code is already testing thoroughly IN/NOT_IN. I added an additional test for the plan

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/LikeCallGen.scala [hunk 1]
```diff
@@ -109,16 +109,16 @@ class LikeCallGen extends CallGenerator {
             val middleMatcher = MIDDLE_PATTERN.matcher(newPattern)
 
             if (noneMatcher.matches()) {
-              val reusePattern = ctx.addReusableStringConstants(newPattern)
+              val reusePattern = ctx.addReusableEscapedStringConstant(newPattern)
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/LikeCallGen.scala:112]: nit: method is quite long `addReusableEscapedStringConstant` -> `addReusableEscapedString`

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/SearchOperatorGen.scala [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+
+    // In case the search is among points we use the hashset implementation
+    if (sarg.isPoints || sarg.isComplementedPoints) {
+      val rangeSet = if (sarg.isPoints) sarg.rangeSet else sarg.rangeSet.complement()
+      val haystack = rangeSet
+        .asRanges()
+        .asScala
+        // We need to go through the generateLiteral to normalize the value from calcite
+        .map(r => toFlinkInternalValue(r.lowerEndpoint, sargType))
+        // The elements are constant, we perform the cast at priori
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/SearchOperatorGen.scala:81]: nit: `a priori` or better `instantly` to help non-native speakers ;)

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/LikeCallGen.scala (hunk 1 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/LikeCallGen.scala:112]: The fact that is adding a constant is quite important, i wouldn't remove it from the name.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/RexLiteralUtil.scala [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      case INTERVAL_DAY_TIME =>
+        value match {
+          case number: JNumber => number.longValue()
+        }
+
+      case DISTINCT_TYPE =>
+        toFlinkInternalValue(value, valueType.asInstanceOf[DistinctType].getSourceType)
+
+      case SYMBOL =>
+        value.asInstanceOf[Enum[_]]
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/RexLiteralUtil.scala:166]: Yep to make the result of this function compatible with `generateLiterals`

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/RexLiteralUtil.scala [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+import java.lang.{Boolean => JBoolean, Integer => JInteger, Number => JNumber}
+import java.math.{BigDecimal => JBigDecimal}
+
+import scala.annotation.tailrec
+
+/**
+ * Utilities to work with [[RexLiteral]]
+ */
+object RexLiteralUtil {
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/RexLiteralUtil.scala]: Reimplemented in Java :)

**Review by twalthr — APPROVED** [review]
Thanks for the update @slinkydeveloper.

**Pull request closed without merging**